docs/source/notes/cuda.rst

```diff
@@ -513,3 +513,189 @@ by GIL of Python interpreter.
+and you suspect its runtime is at least somewhat CPU-limited.
+
+.. _Getting Started with CUDA Graphs:
+    https://developer.nvidia.com/blog/cuda-graphs/
+.. _Graphs section:
+    https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-graphs
+.. _stream capture:
+    https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#creating-a-graph-using-stream-capture
+.. _cudaGraphLaunch:
+    https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1g1accfe1da0c605a577c22d9751a09597
+
+Pytorch API
+^^^^^^^^^^^
+
+.. warning::
+    The API presented here is a prototype and may change in future releases.
+
+Pytorch exposes graphs as a primitive :class:`torch.cuda.CUDAGraph` class,
+which offers :meth:`~torch.cuda.CUDAGraph.capture_begin`,
+:meth:`~torch.cuda.CUDAGraph.capture_end`, and
+:meth:`~torch.cuda.CUDAGraph.replay` methods.
+
+Typically, you shouldn't call :meth:`~torch.cuda.CUDAGraph.capture_begin` or
+:meth:`~torch.cuda.CUDAGraph.capture_begin` yourself. Instead, use one of the
+convenience wrappers (
+:class:`torch.cuda.graph` or
+:class:`torch.cuda.make_graphed_callables`).
+
+:class:`torch.cuda.graph` is a context manager that captures whatever CUDA work
+you run in its context (no more, no less). It's simple and versatile. See
+:ref:`Basic capture<basic-capture>` and
+:ref:`Full-iteration capture<full-iteration-capture>` for examples.
+
+:class:`~torch.cuda.make_graphed_callables` is more sophisticated. It accepts
+Python functions and :class:`torch.nn.Modules`, creates graphs of their forward-pass
+work, and (if ``autograd_aware=True``) also automatically creates graphs of their
+backward-pass work. See
+:ref:`Partial-network capture<partial-network-capture>` for examples.
+
+.. _capture-constraints:
+
+Constraints
+~~~~~~~~~~~
+
+* Capture must occur on a non-default stream. (This is only a concern if you use the raw
+  :meth:`CUDAGraph.capture_begin<torch.cuda.CUDAGraph.capture_begin>` and
+  :meth:`CUDAGraph.capture_end<torch.cuda.CUDAGraph.capture_end>` calls.
+  :class:`~torch.cuda.graph` and
+  :func:`~torch.cuda.make_graphed_callables` set a side stream for you.)
+* Within a process, only one capture may be underway at a time.
+* No non-captured CUDA work may run in this process (on any thread) while capture is underway.
+* CPU work is not captured. If the captured ops include CPU work, that work will be elided during replay.
+* Ops that sychronize the CPU with the GPU (e.g., ``.item()`` calls) are prohibited.
+* Every replay reads from and writes to the same (virtual) memory addresses.
+* Dynamic control flow (based on CPU or GPU data) is prohibited.
+* Dynamic shapes are prohihited. The graph assumes every tensor in the captured op sequence
+  has the same size and layout in every replay.
+* CUDA RNG ops are allowed, but must use default generators. For example, explicitly constructing a
```

Comment:
can we error out on in?

Reply:
I don't think we hard error if a non-default generator is _constructed_ in a capture region, but I think [we do hard error](https://github.com/pytorch/pytorch/blob/773c8b6440eee7d903f3d3a80d1fa866370b1ba2/aten/src/ATen/cuda/CUDAGeneratorImpl.cpp#L272-L274) if an unexpected (non-default) generator is _used_ in an op in a captured region.